lib/managers/KalturaTrackingManager.js

```diff
@@ -0,0 +1,145 @@
+const util = require('util');
+const udpsender = require('../utils/udpsender');
+const crypto = require('crypto');
+const kaltura = require('../KalturaManager');
+
+/**
+ * @service tracking
+ */
+class KalturaTrackingManager extends kaltura.KalturaManager
+{
+    constructor()
+    {
+        super();
+    }
+
+    /***
+     * create a url to call the sendBeacon action when beacon tracking is needed.
+     */
+    static generateBeaconRequest(domain, partnerId, params)
+    {
+
+        KalturaLogger.log('Generating beacon request for domain: [ ' + domain + '] and partnerId [ ' + partnerId + '].');
+        let kalturaManager = new kaltura.KalturaManager();
+        let beaconRequest = kalturaManager.getPlayServerUrl('tracking', 'sendBeacon', partnerId, params, null, domain);
+        KalturaLogger.log('BeaconRequest generated: ' + beaconRequest);
+        return beaconRequest;
+    }
+
+    /***
+     * create a url to call the sendBeacon action when beacon tracking is needed.
+     */
+    static buildTrackingBeaconCacheKey(params)
+    {
+        KalturaLogger.log('Building tracking beacon cache key for partnerId: [' + params.partnerId + '] entryId: [' + params.entryId + '] cuePointId: [' + params.cuePointId+ '] beaconUrl: [' + params.url + ']' + ' headers: [' + JSON.stringify(params.headers) + ']');
+        // build cache key and check for duplicates
+        let keyToHash = params.partnerId + '.' + params.entryId + '.' + params.cuePointId + '.' + params.url + '.' + JSON.stringify(params.headers);
+        let trackingBeaconCacheKey = params.trackingId + "-" + crypto.createHash('md5').update(keyToHash).digest('hex');
+        KalturaLogger.log('Tracking beacon cache key created: ' + trackingBeaconCacheKey);
+        return trackingBeaconCacheKey;
+    }
+
+    /**
+     * Send beacons to track ad progress
+     *
+     * @action tracking.sendBeacon
+     *
+     * @param partnerId
+     * @param entryId
+     * @param cuePointId
+     * @param type
+     * @param url
+     * @param headers
+     */
+    sendBeacon(request, response, params)
+    {
+        response.log('Start sendBeacon.');
+
+        if (!params.url)

```

Comment:
validate all fields - not need for the signature stuff since we can be called from outside